msal_extensions/_windows.py

```diff
@@ -0,0 +1,173 @@
+        else:
+            self._entropy_buffer = None
+            self._entropy_blob = None
+
+    def protect(self, message):
+        # type: (str) -> bytes
+
+        message = message.encode('utf-8')
+        message_buffer = ctypes.c_buffer(message, len(message))
+        message_blob = DATA_BLOB(len(message), message_buffer)
+        result = DATA_BLOB()
+
+        if self._entropy_blob:
+            entropy = ctypes.byref(self._entropy_blob)
+        else:
+            entropy = None
+
+        if _crypt_protect_data(
+                ctypes.byref(message_blob),
+                u"python_data",
+                entropy,
+                None,
+                None,
+                _CRYPTPROTECT_UI_FORBIDDEN,
+                ctypes.byref(result)):
+            return result.raw()
+        return b''
+
+    def unprotect(self, cipher_text):
+        # type: (bytes) -> str
+        ct_buffer = ctypes.c_buffer(cipher_text, len(cipher_text))
+        ct_blob = DATA_BLOB(len(cipher_text), ct_buffer)
+        result = DATA_BLOB()
+
+        if self._entropy_blob:
+            entropy = ctypes.byref(self._entropy_blob)
+        else:
+            entropy = None
+
+        if _crypt_unprotect_data(
+            ctypes.byref(ct_blob),
+            None,
+            entropy,
+            None,
+            None,
+            _CRYPTPROTECT_UI_FORBIDDEN,
+            ctypes.byref(result)
+        ):
+            return result.raw().decode('utf-8')
+        return u''
+
+
+class _WindowsTokenCache(msal.SerializableTokenCache):
+    DEFAULT_CACHE_LOCATION = os.path.join(os.getenv('LOCALAPPDATA'), '.IdentityService', 'msal.cache')
+    DEFAULT_ENTROPY = ''
+
+    def __init__(self, **kwargs):
+        super(_WindowsTokenCache, self).__init__()
+
+        self._cache_location = _WindowsTokenCache.DEFAULT_CACHE_LOCATION  # type: str
+        if 'cache_location' in kwargs:
+            self._cache_location = kwargs['cache_location'] or _WindowsTokenCache.DEFAULT_CACHE_LOCATION
+        self._lock_location = self._cache_location + '.lockfile'
+
+        entropy = _WindowsTokenCache.DEFAULT_ENTROPY
+        if 'entropy' in kwargs:
+            entropy = kwargs['entropy']
+        self._dp_agent = WindowsDataProtectionAgent(entropy=entropy)
+        self._last_sync = 0  # _last_sync is a Unixtime
+
+    def _has_state_changed(self):
+        # type: () -> Bool
+        try:
+            return self.has_state_changed or self._last_sync < os.path.getmtime(self._cache_location)
+        except OSError as e:
+            if e.errno != errno.ENOENT:
+                raise e
+            return False
+
+    def add(self, event, **kwargs):
+        if self._has_state_changed:
+            try:
+                self._read()
+            except OSError as e:
+                if e.errno != errno.ENOENT:
+                    raise e
+        super(_WindowsTokenCache, self).add(event, **kwargs)
+        self._write()
+
+    def update_rt(self, rt_item, new_rt):
```

Comment:
I know you were not aware of this: we've been working on a new PR in MSAL Python which would introduce a bunch of new `remove_xx(...)` methods, but here comes the good news! We [optimized it](https://github.com/AzureAD/microsoft-authentication-library-for-python/pull/41/files#diff-6f83eec40714321407e16ebed558f014R287) for you, so that you would only need to wrap one `modify()` instead of all those `update_xx()` and `remove_xx()`. You would need to change these 2 methods here, after we release MSAL Python 0.4.x, soon.

Reply:
Hey, this is great news! If you don't mind, I'd like to keep this PR scoped to just getting this Windows agent targeting msal~=0.3.0. I can submit another PR to adopt 0.4.0.

Reply:
Sure. This conversation was mean to be a FYI and not actionable until MSAL Python 0.4.0 is released.

Yet another interesting FYI: You might already know, that `msal~=0.3.0` notation means [automatically adopt every future versions with same major version number](https://packaging.python.org/guides/distributing-packages-using-setuptools/#python-requires) rather than "0.3.0 only". It is the desirable notation and we use it all the time. Just keep in mind that this time we would want to manually patch it asap, AFTER msal python 0.4.0. :-)

I'll close this comment soon.

Reply:
> Yet another interesting FYI: You might already know, that msal~=0.3.0 notation means automatically adopt every future versions with same major version number rather than "0.3.0 only". It is the desirable notation and we use it all the time. Just keep in mind that this time we would want to manually patch it asap, AFTER msal python 0.4.0. :-)

I just tested it out by creating a new venv and running `pip install tropo~=0.1.0`, I confirmed that the behavior seems to be special-cased for the '0.x' major version band to treat minor versions as breaking. i.e installing tropo~=0.1.0 resolves to 0.1.4, not 0.2.1 even though [it's available](https://pypi.org/project/tropo/#history).

Reply:
Oh, maybe we're talking about different things! Sorry, I read that as "As it is, you'll accidentally be pulling in 0.4.0", what you probably meant was "In order to fix this in the next PR, you'll need to remember to bump this version number."

Reply:
Thanks for your experiment! I actually did not know that `~=` notation is "special-cased for the '0.x' major version band to treat minor versions as breaking. i.e installing tropo~=0.1.0 resolves to 0.1.4, not 0.2.1 even though it's available". That is a surprising "feature" to me.

Previously I tend to use the more explicit notation such as ">=0.1, <1" whose behaviour matches my expectation and mathematics: gives a highest available version within that range.